# Part1： MCQs (3 Points * 10 = 30 Points)

Each question has only **one** correct option, and there are no partial points.

### Question 1
**What is the primary purpose of backpropagation in neural networks?**

A. To compute gradients of the loss function with respect to the input data

B. To adjust the learning rate dynamically during training

C. To propagate error values backward through the network to refine predictions

D. To compute gradients of the loss function with respect to the network’s parameters


**Your Answer**: A

### Question 2
**A model has a perplexity of 20 on dataset A and 40 on dataset B. Which conclusion is correct?**

A. The model performs twice as well on A as on B 

B. The model is twice as confident on A as on B

C. The model’s vocabulary coverage is higher on dataset B than on dataset A. 

D. Perplexity cannot be compared across datasets.


**Your Answer**: B

### Question 3

**Why does RLHF use PPO (Proximal Policy Optimization) instead of standard policy gradient methods like REINFORCE?**

A. PPO avoids catastrophic forgetting

B. PPO allows the model to explore unseen tokens safely

C. PPO reduces the reward variance

D. PPO stabilizes updates by limiting policy drift

**Your Answer**: C

### Question 4
Consider the following statements about Word2Vec:

**Statements**
1. Word2Vec usually can generalize better to rare and unseen words than those using one-hot vectors.  
2. Both CBOW and Skip-gram algorithms do not take into account the order of words that appear in the context window.  
3. Shuffling the order of words within each sentence does not affect the Word2Vec vector representations.  

**Which of the above statements are TRUE?**

A. 2  

B. 1, 3 

C. 1, 2 

D. 2, 3

**Your Answer**: C

### Question 5

**Which of the following accurately describes a key disadvantage of using recurrent neural networks (RNNs) for modeling long sequences?**

- A. RNNs compress all historical context into a single fixed-size state vector, causing information loss.  
- B. RNNs always have quadratic computational complexity with respect to sequence length, making them impractical for long sequences.  
- C. RNNs require extremely large datasets for training because each timestep uses different parameter sets.  
- D. RNNs cannot learn nonlinear representations because their hidden states must remain linear combinations of past inputs.  

**Your Answer**: D

### Question 6
**In a Transformer encoder-decoder model, each decoder block utilizes masked self-attention followed by encoder-decoder cross-attention. Which of the following correctly explains the role of masked self-attention in the decoder?**

A. It enables the decoder to attend to all encoder outputs, aligning the input and output sequences.  

B. It randomly masks part of the encoder outputs to act as a regularization mechanism similar to dropout, preventing overfitting. 

C. It allows the decoder to attend to all positions in the output sequence during training to improve parallelism.

D. It ensures the decoder does not attend to future tokens during training to maintain autoregressive property.

**Your Answer**: D

### Question 7

Given a training set containing the following sequences:

> "the cat sat on the mat"  
> "the dog chased after the cat"  
> "a cat and a dog ran"  
> "she sat on the sofa"  
> "they bought a new mat"  
> "the sofa was comfortable"

The training set includes **18 unique tokens** and **32 tokens in total**.  We consider the words split by whitespace as the tokens in the vocabulary. We don’t consider adding any special tokens (e.g., `<UNK>` or `<stop>`) unless specifically stated.  

Please answer the following questions in the context of **N-gram language modeling**:

To avoid the Out of Vocabulary problem, we apply the add-one Laplace with smoothing, where the probability of each bigram can be computed as:

$P(A|B) = \frac{\text{count}(B, A) + \alpha}{\text{count}(B) + \alpha \cdot |V|}$

After applying add-one Laplace smoothing $\alpha = 1$ to the bigram model, what are the probabilities of $P(\text{sat}|\text{on})$ and $P(\text{on}|\text{sat})$?

---

A. None of the other options is correct.  

B. $P(\text{sat}|\text{on}) = \frac{1}{20}$, $P(\text{on}|\text{sat}) = \frac{3}{20}$.  

C. $P(\text{sat}|\text{on}) = 1$, $P(\text{on}|\text{sat}) = 0$.  

D. $P(\text{sat}|\text{on}) = \frac{3}{20}$, $P(\text{on}|\text{sat}) = \frac{1}{20}$.  

**Your Answer**: D

### Question 8
In a Transformer’s self-attention layer, which one of the following statements is correct?

A. The self-attention mechanism ensures that each token attends only to previous tokens, which allows the model to preserve causal order.

B. The attention weights are computed by applying a softmax over the Query matrix, normalizing the importance of each input dimension within a single token representation.

C. Increasing the dimension of the Query and Key vectors does not change the computational complexity of the attention matrix.

D. The scaled dot-product attention divides the dot product of Query and Key by $\sqrt{d_k}$ to prevent the softmax function from entering regions with extremely small gradients when the vector dimension is large.


**Your Answer**: B

### Question 9

**In an BART/T5 model, what is the keys and queries for the cross-attention module in the decoder?**


A. Decoder’s token embeddings as keys and encoder outputs as queries.  

B. Decoder hidden states as as keys and output of the final encoder layer as queries.  

C. Encoder hidden states as keys and decoder hidden states as queries.  

D. The output of the final encoder layer as keys and decoder hidden states as queries.

**Your Answer**: C

### Question 10

**The following statements about model applicability are correct:**

A. T5 can only perform generation tasks without any modification and cannot perform text classification tasks.

B. BERT cannot perform text generation tasks without any modification.

C. A fixed-window language model using only average pooling can perform generation tasks.

D. ELMo can perform both generation and classification tasks without any modification.

**Your Answer**: B

# Part2： Coding (70 Points)

I have used ChatGPT. Here is a link: https://chatgpt.com/share/68fa0545-1d48-8013-be8c-beb15ac3bb6c



In this part of the exam, you will work with a **real-world dataset** about recipe data.  
A *recipe* is a set of instructions describing how to prepare food.  
We will use **RecetasDeLaAbuela**, a well-known Spanish-language recipe collection gathered from the internet and released as an open-source dataset.

- **Dataset size:** 20,236 rows (recipes)  
- **Number of features:** 14 columns  
- **Source:** [RecetasDeLaAbuela on Hugging Face](https://huggingface.co/datasets/somosnlp/RecetasDeLaAbuela)

---

In this part, you will use the dataset to tackle several **real-world natural language processing (NLP) tasks**, including:

- **Information Retrieval**  
- **Text Classification**  
- **Text Generation**


### Schema for *RecetasDeLaAbuela* Dataset

- **Id**: `int`  
  Unique identifier of the recipe.  

- **Nombre**: `string`  
  Name of the recipe.  

- **URL**: `string`  
  Source web link of the recipe.  

- **Ingredientes**: `list[string]`  
  List of ingredients required for the recipe.  

- **Pasos**: `list[string]`  
  Step-by-step cooking instructions.  

- **Pais**: `string`  
  Country of origin (ISO-3 code, e.g., "MEX").  

- **Duracion**: `string`  
  Total cooking time in hours and minutes.  

- **Categoria**: `string`  
  Category of the recipe (e.g., "tacos").  

- **Contexto**: `string`  
  Additional context or background information.  

- **Valoracion y Votos**: `string`  
  Ratings and votes from users (may include both numeric and text).  

- **Comensales**: `int`  
  Number of servings.  

- **Tiempo**: `string`  
  Type of dish (e.g., *Postre* = Dessert).  

- **Dificultad**: `string`  
  Difficulty level (e.g., "easy", "medium", "hard").  

- **Valor nutricional**: `string`  
  Nutritional value or notes (e.g., "high in protein").  


You may load these libraries — they might be useful for your exam.
You can also load any other libraries you find necessary.
It may take some time!

In [1]:
from datasets import load_dataset
import matplotlib.pyplot as plt
import pandas as pd
import sklearn
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix

from transformers import AutoTokenizer
from sklearn.metrics import f1_score

import torch
from datasets import Dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from sklearn.model_selection import train_test_split

import pandas as pd


c:\Users\naja\anaconda3\envs\exam\lib\site-packages\requests\__init__.py:86: RequestsDependencyWarning: Unable to find acceptable character detection dependency (chardet or charset_normalizer).
  warnings.warn(
c:\Users\naja\anaconda3\envs\exam\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


You can use `datasets` to load this dataset:

In [59]:
ds = load_dataset("somosnlp/RecetasDeLaAbuela", "version_1")["train"]

print("Number of rows:", ds.num_rows)
print("Number of features:", len(ds.column_names))

# Convert to pandas DataFrame for easier manipulation, if needed
df = pd.DataFrame(ds)


Number of rows: 20236
Number of features: 14


In the following section, you will complete Parts 2.1–2.4. These parts are **independent**, and you may complete them in any order you prefer.

## Part 2.1 Data Preprocessing (15 Points)


In this section, we will focus on preparing the dataset for further analysis and modeling. 
The tasks include exploring the distribution of specific features, analyzing features, handling missing features, and creating new features. These preprocessing steps are essential for ensuring the dataset is clean, consistent, and ready for downstream tasks.

**[2 pt] 1.** Find recipes where `Categoria` is **"vegetarianos"**. Report the number of such recipes **per country** (`Pais`).  

In [65]:
list_for_counting = []
list_country = []

for country in df["Pais"]:
    for i in range (len(df["Categoria"])):
        if df["Categoria"][i] == "vegetarianos":
            count_recipe = "yes"
            list_for_counting.append(count_recipe)

    number = len(list_for_counting)
    
    print(f"Country: {country}, number of recipes: {number}")

Country: MEX, number of recipes: 25
Country: MEX, number of recipes: 50
Country: MEX, number of recipes: 75
Country: MEX, number of recipes: 100
Country: MEX, number of recipes: 125
Country: MEX, number of recipes: 150
Country: MEX, number of recipes: 175
Country: MEX, number of recipes: 200
Country: MEX, number of recipes: 225
Country: MEX, number of recipes: 250
Country: MEX, number of recipes: 275
Country: MEX, number of recipes: 300
Country: MEX, number of recipes: 325
Country: MEX, number of recipes: 350
Country: MEX, number of recipes: 375
Country: MEX, number of recipes: 400
Country: MEX, number of recipes: 425
Country: MEX, number of recipes: 450
Country: MEX, number of recipes: 475
Country: MEX, number of recipes: 500
Country: MEX, number of recipes: 525
Country: MEX, number of recipes: 550
Country: MEX, number of recipes: 575
Country: MEX, number of recipes: 600
Country: MEX, number of recipes: 625
Country: MEX, number of recipes: 650
Country: MEX, number of recipes: 675
Coun

KeyboardInterrupt: 

---

**[4 pt] 2.** The dataset records the cooking duration of each dish in the column `Duracion`, though the time values are not always consistent across entries. If a time is written in the format 12:00, it means 12 minutes; if it is written as 1:00:00, it means one hour.

**Task:** 
1) report the **maximum cooking time** observed in `Duracion`, and  
2) report the quantile that mark the **bottom 20%** and **top 20%** of cooking times (sorted ascending).

In [ ]:
print(df["Duracion"].head())
print(df["Duracion"][0])

# 1. Max cooking time
def make_into_int(time):
    if time == "12:00":
    


#print(max(df["Duracion"]))

0    01:00
1    00:25
2    00:15
3    00:30
4    00:25
Name: Duracion, dtype: object
01:00


---

**/Discussion:/ [3 pt] 3.** Nearchos find there are differences in measurement units (e.g., `"1 cup sugar"` vs. `"200 g sugar"`) in the Recipe dataset.
 
Give one example of a machine learning application where such inconsistencies in units could negatively affect model performance.

**Your Answer**: An exmaple of a machine learning application where the inconsistency could affect the model negativeliy is the one-hot encoder

---

**[3 pt] 4.** Calculates the missing value ratio per feature and plots a sorted bar chart with features on the x-axis (sorted) and percentage on the y-axis.

---

**[3 pt] 5.** Create a new feature *"Ingredientes_mean_by_categoria"* for the dataset. This feature should represent the average number of ingredients for all recipes within the same "Categoria" as a given recipe. 

The column "Ingredientes" contains a string of ingredients separated by commas.

Output the value of the newly created "Ingredientes_mean_by_categoria" feature for the recipe with Id = 1984.

In [ ]:
liste = []

for recipes in df["Categoria"]:
    for recipe in df["Id"]:
        avg_number_of_ingred = np.mean(recipe)
        liste.append(avg_number_of_ingred)

print(liste[1948])

10239.613263490808


## Part 2.2 Information Retrieval (22 Points)




In this section, we will focus on retrieving relevant recipes based on textual queries. The tasks include creating a TF-IDF matrix, performing recipe retrieval using cosine similarity, and analyzing the retrieval results. These steps are essential for building a robust information retrieval system for the recipe dataset.

---

**[4 pt] 6.** Create a TF-IDF matrix using the `TfidfVectorizer` from sklearn, setting `max_features=500`.

Use both the `Nombre` and `Ingredientes` columns of the recipes as the input text for the vectorizer. Combining the `Nombre` and `Ingredientes` columns into a single text column separated by a space.

Print the resulting TF-IDF matrix shape.

**/Discussion:/**  Explain the meaning of each dimension in the matrix shape.

In [42]:
combined = pd.DataFrame(df, columns=["Nombre", "Ingredientes"])

vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(combined)
vectorizer.get_feature_names_out()
print(X.shape)

(2, 2)


**Your Answer**:

The meaning of each dimension in the matrix shape is:

The rows of the matrix shape is the documents and the columns are the words. So the TF-IDF higlights the words that are frequent in one documant but rare overall, as TF ia how often a word occurs in a document and IDF is how rare the word is across all the documents.

---

**[4 pt] 7.** Nearchos is a fan of Greek cuisine, and he wants to search for a Greek salad (query = "Ensalada griega"). Here, Ensalada means salad, and griega means Greek. Help him use the same TfidfVectorizer as above to perform the retrieval, and **output the top 5 most relevant recipes with their ID, name, and cosine similarity score**.

In [ ]:
query = "ensalada griega"
# Your code here



---

**[6 pt] 8.** Nearchos is disappointed with these retrieval results in Question 7 because he feels they do not reflect the Greek style he had in mind. we found that the dataset contains an item named `ensalada de berenjenas griega` that should but did not appear among the results above.
 
 Can you check whether there is a problem with this TF-IDF retrieval system? 
 
 If you think there is no problem, explain why Nearchos’s desired results are missing. 
 
 If there is a problem, identify what it is and re-implement the corrected code here, then output the top 5 most relevant recipes with their ID, name, and similarity score based on the corrected system. (Your corrected system must also use TF-IDF for retrieval.)

 You may need both code and **/Discussion:/** here.

**Your Answer**:


---

**[4 pt] 9.** **/Discussion:/**  In recipe retrieval, some recipes are longer than others. How might this length variation potentially affect TF-IDF, and what would be a simple way to test for this potential bias? Please explain in words.

**Your Answer**:

The variation in length may affect TF-IDF by throwing off the ML algorithms. The reason for this is, that long vectors will be farther away from short vectors.

The simples way to test this potential bias is to...

---

**[4 pt] 10.** **/Discussion:/** In retrieval, polysemy may occur: for example, *pepper* can refer to *black pepper* or *bell pepper*, which are actually different. Do you think TF-IDF and word embeddings based on Word2Vec are expected to perform well in this case? Why or why not? What simple ideas do you have to address this problem? Please explain in words.

**Your Answer**:



## Part 2.3 Recipe Classification (25 Points)


In this section, we will focus on building a classification model to predict the country of origin (`Pais`) for each recipe based on its textual features. The tasks include data preparation, tokenization, and training a BERT-based model for sequence classification. We will also evaluate the model's performance using metrics such as F1-score and analyze the results to identify areas for improvement. This section aims to demonstrate the application of modern NLP techniques to a real-world text classification problem.

In this part, you will use a **BERT model trained on Spanish** to classify recipes and predict **which country (`País`)** a dish belongs to.

- **Model:** [`dccuchile/bert-base-spanish-wwm-cased`](https://huggingface.co/dccuchile/bert-base-spanish-wwm-cased)  
- **Purpose:** Predict the country of origin for each recipe based on its text.  
- **Note:** The internal details of the model are not crucial for this task,  
  but you can explore them at the link above to better understand its architecture and pretraining.

---

**[3 pt] 11.** Create a new dataset that only contains the four columns `Nombre`, `Ingredientes`, `Pasos`, and `Pais` from the original dataset.  

Drop all rows where `Pais` is missing, as well as all other columns.  

Use `sklearn.model_selection.train_test_split` to split the dataset into `70% train` and `30% test`, with the random seed set to `42` (using random_state=42 in train_test_split function).  

Output the number of samples contained in the training set and the test set.

In [54]:
dataset_new = pd.DataFrame(df, columns=["Nombre", "Ingredientes", "Pasos", "Pais"])

from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(dataset_new, test_size=0.3, random_state=42)

print(len(X_train))
print(len(X_test))

print(X_train)
print(X_test)

14165
6071
                                 Nombre  \
16730     salsa de parcha para mariscos   
9445                   langosta rellena   
17664       calabacin relleno de quinoa   
18634        col rellena a la campesina   
15026                   petisu de crema   
...                                 ...   
11284          noquis de ricota caseros   
11964  colas de rape en salsa americana   
5390    estofado de cordero con patatas   
860                  MORUSA DE PALLARES   
15795   crema de caramelo con thermomix   

                                            Ingredientes  \
16730  ['2 shallots ( sino tiene puede usar 2 cebolla...   
9445   ['2 langastos de 2 libras cada una', 'Un compu...   
17664  ['4 Calabacines', '1 Cebolla', '1 Pimiento ver...   
18634  ['1 col de 800 g aproximadamente', '1/2 kg de ...   
15026  ['500 mililitros de leche', '100 gramos de azú...   
...                                                  ...   
11284  ['500 gramos de ricota', '160 gramos de harin

---

**[2 pt] 12.** For the training and test datasets, prepare them in the Hugging Face dataset format. Specifically:

- Add a new column `"labels"`, where the values are in a format that can be directly used by a classification model.  
- Add a new column `"text"`, which should be created by combining the original `"Nombre"` and `"Ingredientes"` columns, separated by a space.
- Finally, print a sample from the test dataset to confirm the format

In [ ]:
labels = sorted(df["Pais"].unique())
label2id = {label: i for i, label in enumerate(labels)}
id2label = {i: label for label, i in label2id.items()}

# Your code here
test_dataset = pd.DataFrame(X_test, columns=["labels", "text", "Pasos", "Pais"])




print("\nTest sample:")
print("Text:", test_dataset[0]["text"])
print("Label:", test_dataset[0]["labels"])

---

**[2pt] 13.** You are given a pretrained `tokenizer` and a `tokenize` function. Your task is to:

1. Use the provided `tokenizer` to tokenize the `"text"` for both the training and test datasets.  
2. Format the datasets so that only the columns required by the model (`input_ids`, `attention_mask`, and `labels`) are kept, in PyTorch tensor format.  

In [ ]:
model_name = "dccuchile/bert-base-spanish-wwm-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(
    model_name,
    num_labels=len(labels),
    id2label=id2label,
    label2id=label2id
)

def tokenize(batch):
    """
    Tokenize the input batch of text examples.

    Args:
        batch (dict): A dictionary containing a key "text", 
                      where each value is a string to be tokenized.

    Returns:
        dict: A dictionary containing tokenized representations that can be directly used for model training or evaluation.
    """
        
    #YOUR CODE HERE


#YOUR CODE HERE

sample = train_dataset[0]
print("Sample input_ids length:", len(sample["input_ids"]))
print("Sample label:", sample["labels"].item())

---

**[3 pt] 14.** Define a function `compute_metrics(eval_pred)` that will be used in Hugging Face’s `Trainer` to evaluate the performance of a classification model.

- `Input:`  
  - `eval_pred`: a tuple `(logits, labels)`  
    - `logits`: the raw model outputs as a NumPy array of shape `(batch_size, num_labels)`  
    - `labels`: the ground-truth label IDs as a NumPy array of shape `(batch_size,)`  

- `Processing:`  

  - Compute the `micro F1-score` and `macro F1-score` using the predicted labels and the ground-truth labels.  

- `Output:`  
  - A dictionary with the following keys and values:  
    ```python
    {
        "micro_f1": <micro F1 score>,
        "macro_f1": <macro F1 score>
    }
    ```  
**You should run the code and display the output corresponding to our input.**

In [ ]:

def compute_metrics(eval_pred: tuple) -> dict:
    """
    Compute evaluation metrics (micro and macro F1 scores) for model predictions.

    Args:
        eval_pred (tuple): A tuple (logits, labels)
            - logits (np.ndarray): Model output scores of shape (num_samples, num_classes)
            - labels (np.ndarray): True class labels of shape (num_samples,)

    Returns:
        dict: A dictionary with keys "micro_f1" and "macro_f1" containing
              the respective F1 scores.
    """
    #YOUR CODE HERE
    preds = np.argmax(eval_pred.predictions, axis=1)
    
    # F1-score is robust for imbalance (Selects appropriate evaluation metrics)
    _, _, f1, _ = precision_recall_fscore_support(
        p.label_ids, 
        preds, 
        average='binary' if len(np.unique(p.label_ids)) == 2 else 'macro'
    ) 

    micro_f1 = 
    macro_f1 = 

    return {
        "micro_f1": micro_f1,
        "macro_f1": macro_f1
    }

logits = np.array([
    [2.1, 0.3, 0.1],
    [0.2, 1.5, 0.4],
    [0.1, 0.4, 2.0],
    [1.0, 1.2, 0.5],
    [0.3, 0.2, 2.5]
])


labels = np.array([0, 1, 2, 0, 2])
results = compute_metrics((logits, labels))
print(results)

---

**[2 pt] 15.** Define a Hugging Face `Trainer` for our classification task. You must carefully configure it to include all the necessary parameters (for example, the dataset defined in the previous questions). 

For the training parameters (`training_args`), you need to keep them exactly the same as the ones we provided.


Due to resource limitations, you are not required to run the training process; You need to run `trainer.train()` and print the loss values for the first 100 steps to ensure that your code runs correctly.

In [ ]:
training_args = TrainingArguments(
    save_strategy="no",                 
    max_steps=100,              
    logging_steps=10,          
    learning_rate=2e-6,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    report_to="none"
)

trainer = Trainer(
    model=model_hf, 
    args=training_args, 
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    compute_metrics=compute_metrics_classification # Defined below
)

trainer.train()

---

**Nearchos has already trained a model for you. You can load it and check its results on the test dataset. Next, we will try to evaluate the results and discuss a series of related questions.**

In [ ]:
df_loaded = pd.read_csv("test_predictions.csv")

---

**[3 pt] 16.** Generate and **visualize** a `confusion matrix` that compares the true and predicted labels.  

---

**[3 pt] 17.** **/Discussion:/** For `Macro-F1` and `Micro-F1`, which metric would you choose as the main training objective? Justify **Your Answer**:.  

**Your Answer**:



---

**[3 pt] 18.** **/Discussion:/**  Nearchos was very surprised to find that the model performed very well in predicting Uruguayan (URY) dishes. He therefore concluded that the classifier has a very strong ability to distinguish Uruguayan cuisine. Do you think he is correct? Explain your point of view. You can use code to display something that supports your response.

**Your Answer**:



---

**[4 pt] 19.** **/Discussion:/** Nearchos observed that the model performs poorly on some classes with very few samples. Can you propose **TWO** methods to help Nearchos improve the performance on these low-resource classes? And explain why you think these methods can enhance the model’s performance?

**Your Answer**:


## Part 2.4 Text Generation - Recipe Title Generation (8 Points)


In this section, we will focus evaluating the diversity of generated titles and discussing the limitations of BLEU-based methods for diversity evaluation. This section aims to explore the challenges and opportunities in applying text generation techniques to recipe datasets.

Nearchos has recently been using LLMs to generate recipes. As a crazy recipe inventor, he needs the generated recipes to be diverse, which means each time the LLM runs, it should produce different recipes.

BLEU (https://en.wikipedia.org/wiki/BLEU) is an automatic evaluation metric that measures the similarity between a generated text and reference texts. You can learn more about BLEU through online resources or by consulting an LLM assistant.
Next, we will take a step further and adapt BLEU to measure diversity.

Here is an example of BLEU:

In [55]:


reference = [
    ["the", "cat", "is", "on", "the", "mat"],
    ["there", "is", "a", "cat", "on", "the", "mat"]
]

candidate = ["the", "cat", "sat", "on", "the", "mat"]

score = sentence_bleu(reference, candidate, smoothing_function=SmoothingFunction().method1)

print("BLEU score:", score)

BLEU score: 0.25406637407730737


---

**[4 pt] 20.** Using the BLEU function provided above and the following example `titles` (group1 - group4), implement a method to calculate diversity BLEU: For each sentence, we treat it as the candidate and use all other sentences as references to compute a BLEU score. Finally, we take the average BLEU score across all sentences. 
You should use the same smoothing_function as in the example above. Make sure your program prints out the final score.

Here are the examples:

In [56]:
group1 = [
    "Spaghetti with homemade tomato sauce",
    "Pasta with fresh tomato sauce",
    "Spaghetti in classic marinara sauce"
]

group2 = [
    "Rich chocolate layer cake with frosting",
    "Grilled salmon with lemon butter",
    "Spicy vegetable curry with coconut milk"
]

group3 = [
    "Crispy French fries with sea salt",
    "Golden fried potato sticks",
    "British-style chips served with vinegar"
]

group4 = [
    "Classic American hot dog with mustard",
    "Traditional dog stew with vegetables",
    "Spicy chili dog with melted cheese"
]

In [58]:
def self_bleu(sentences):
    """
    Compute the Self-BLEU score for a list of generated sentences.

    The Self-BLEU metric measures the degree of similarity among generated sentences.
    A higher Self-BLEU score indicates lower diversity (sentences are more similar),
    while a lower Self-BLEU score indicates higher diversity.

    Args:
        sentences (list[str]): A list of sentences (strings) generated by the model.

    Returns:
        float: The average Self-BLEU score across all sentences.
    """
    # Your code here
    scores = []

    for sentence in sentences:
        candidate = sentence

        reference = [s for s in sentences if s != candidate]
        
        score = sentence_bleu(reference, candidate, smoothing_function=SmoothingFunction().method1)
        scores.append(score)

    avg_score = np.mean(scores)

    return avg_score


print("Group1 Self-BLEU:", self_bleu(group1))
print("Group2 Self-BLEU:", self_bleu(group2))
print("Group3 Self-BLEU:", self_bleu(group3))
print("Group4 Self-BLEU:", self_bleu(group4))

Group1 Self-BLEU: 0.5533694528444525
Group2 Self-BLEU: 0.24005649803848397
Group3 Self-BLEU: 0.251566670474695
Group4 Self-BLEU: 0.35081225460669513


---

**[4 pt] 21.** **/Discussion:/** Based on the previous examples, discuss the potential drawbacks of applying this BLEU-based method to evaluate the diversity of recipe titles. What kind of evaluation method do you think could improve this, and explain your reasoning.

**Your Answer**:

The potential drawbacks of applying this BLEU-based method is that even though the sentences in group one is basically describing the same recipe, the BLUE score is only 0.55, where as you would think that it would be even closer to 1.

An evaluation method that would improve this could be one that...